packages/ui/src/hooks/_utils/responsiveObserve.ts

```diff
@@ -0,0 +1,74 @@
+export type Breakpoint = 'mobileSmall' | 'mobileLarge' | 'tabletSmall' | 'tabletLarge' | 'desktopSmall' | 'desktopLarge'
+export type BreakpointMap = Record<Breakpoint, string>
+export type ScreenMap = Partial<Record<Breakpoint, boolean>>
+export type ScreenSizeMap = Partial<Record<Breakpoint, number>>
+
+export const responsiveArray: Breakpoint[] = ['mobileSmall', 'mobileLarge', 'tabletSmall', 'tabletLarge', 'desktopSmall', 'desktopLarge']
+
+export const responsiveMap: BreakpointMap = {
+    mobileSmall: '(max-width: 480px)',
+    mobileLarge: '(min-width: 480px)',
+    tabletSmall: '(min-width: 768px)',
+    tabletLarge: '(min-width: 992px)',
+    desktopSmall: '(min-width: 1200px)',
+    desktopLarge: '(min-width: 1600px)',
+}
+
+type SubscribeFunc = (screens: ScreenMap) => void
+const subscribers = new Map<number, SubscribeFunc>()
+let subUid = -1
+let screens = {}
+
+const responsiveObserve = {
+    matchHandlers: {} as {
+        [prop: string]: {
+            mql: MediaQueryList;
+            listener: ((this: MediaQueryList, ev: MediaQueryListEvent) => any) | null;
+        };
+    },
+    dispatch (pointMap: ScreenMap) {
+        screens = pointMap
+        subscribers.forEach(func => func(screens))
+        return subscribers.size >= 1
+    },
+    subscribe (func: SubscribeFunc): number {
+        if (!subscribers.size) this.register()
+        subUid += 1
+        subscribers.set(subUid, func)
+        func(screens)
+        return subUid
+    },
+    unsubscribe (token: number) {
+        subscribers.delete(token)
+        if (!subscribers.size) this.unregister()
+    },
+    unregister () {
+        Object.keys(responsiveMap).forEach((screen) => {
+            const matchMediaQuery = responsiveMap[screen as Breakpoint]
+            const handler = this.matchHandlers[matchMediaQuery]
+            handler?.mql.removeListener(handler?.listener)
+        })
+        subscribers.clear()
+    },
+    register () {
+        Object.keys(responsiveMap).forEach((screen) => {
+            const matchMediaQuery = responsiveMap[screen as Breakpoint]
+            const listener = ({ matches }: { matches: boolean }) => {
+                this.dispatch({
+                    ...screens,
+                    [screen]: matches,
+                })
+            }
+            const mql = window.matchMedia(matchMediaQuery)
```

Comment:
Can you test it on a simple NextJS server or something like that...not sure about SSR environment.

I'm sure, that calling `document` directly causes issues, but not sure about window

Reply:
`ResponsiveObserve` called only in `useEffect` in `useBreakpoint` hook.
This `useEffect` will not be called on ssr because it is independent of the data received on the server.

But just in case, add a window check in `useEffect`.